substrate/client/network/sync/src/strategy/state.rs

```diff
@@ -0,0 +1,758 @@
+			error!(
+				target: LOG_TARGET,
+				"Failed to downcast opaque state response, this is an implementation bug."
+			);
+
+			BadPeer(peer_id, rep::BAD_RESPONSE)
+		})?;
+
+		debug!(
+			target: LOG_TARGET,
+			"Importing state data from {} with {} keys, {} proof nodes.",
+			peer_id,
+			response.entries.len(),
+			response.proof.len(),
+		);
+
+		match self.state_sync.import(*response) {
+			ImportResult::Import(hash, header, state, body, justifications) => {
+				let origin = BlockOrigin::NetworkInitialSync;
+				let block = IncomingBlock {
+					hash,
+					header: Some(header),
+					body,
+					indexed_body: None,
+					justifications,
+					origin: None,
+					allow_missing_state: true,
+					import_existing: true,
+					skip_execution: true,
+					state: Some(state),
+				};
+				debug!(target: LOG_TARGET, "State download is complete. Import is queued");
+				self.actions
+					.push(StateStrategyAction::ImportBlocks { origin, blocks: vec![block] });
+				Ok(())
+			},
+			ImportResult::Continue => Ok(()),
+			ImportResult::BadResponse => {
+				debug!(target: LOG_TARGET, "Bad state data received from {peer_id}");
+				Err(BadPeer(peer_id, rep::BAD_STATE))
+			},
+		}
+	}
+
+	/// A batch of blocks have been processed, with or without errors.
+	///
+	/// Normally this should be called when target block with state is imported.
+	pub fn on_blocks_processed(
+		&mut self,
+		imported: usize,
+		count: usize,
+		results: Vec<(Result<BlockImportStatus<NumberFor<B>>, BlockImportError>, B::Hash)>,
+	) {
+		trace!(target: LOG_TARGET, "State sync: imported {imported} of {count}.");
+
+		let results = results
+			.into_iter()
+			.filter_map(|(result, hash)| {
+				if hash == self.state_sync.target_hash() {
+					Some(result)
+				} else {
+					debug!(
+						target: LOG_TARGET,
+						"Unexpected block processed: {hash} with result {result:?}.",
+					);
+					None
+				}
+			})
+			.collect::<Vec<_>>();
+
+		if !results.is_empty() {
+			// We processed the target block
+			results.iter().filter_map(|result| result.as_ref().err()).for_each(|e| {
+				error!(
+					target: LOG_TARGET,
+					"Failed to import target block with state: {e:?}."
+				);
+			});
+			match results.into_iter().any(|result| result.is_ok()) {
+				true => {
+					info!(
+						target: LOG_TARGET,
+						"State sync is complete ({} MiB), continuing with block sync.",
+						self.state_sync.progress().size / (1024 * 1024),
+					);
+				},
+				false => {
+					error!(
+						target: LOG_TARGET,
+						"State sync failed. Falling back to full sync.",
+					);
+				},
+			}
+			self.actions.push(StateStrategyAction::Finished);
+		}
+	}
+
+	/// Produce state request.
+	fn state_request(&mut self) -> Option<(PeerId, OpaqueStateRequest)> {
+		if self.state_sync.is_complete() {
+			return None
+		}
+
+		if self
+			.peers
+			.iter()
+			.any(|(_, peer)| matches!(peer.state, PeerState::DownloadingState))
```

Comment:
nit: how about?
`if !self.peers.values().all(Peer::is_available)`

Reply:
I'd like to keep the check for `PeerState::DownloadingState` for consistency with other strategies, as other peer states can be allowed should we introduce them in the future. But thanks for the hint that we can use `values()` instead of `iter()` here.

Reply:
`any(Peer::is_downloading)` would also be nice. it is minor however, up to you...

Reply:
So far have decided to keep `matches!` to not introduce the `is_...()` for every enum variant...